File: src/main/java/essentials/feature/Permission.java
package essentials.feature;

import arc.Core;
import essentials.core.player.PlayerData;
import essentials.internal.Bundle;
import essentials.internal.CrashReport;
import essentials.internal.Log;
import essentials.internal.exception.PluginException;
import mindustry.entities.type.Player;
import org.hjson.*;
import org.slf4j.LoggerFactory;

import java.io.IOException;

import static essentials.Main.*;
import static mindustry.Vars.playerGroup;
import static org.hjson.JsonValue.readHjson;

public class Permission {
    public JsonObject permission;
    public JsonObject permission_user;
    public String default_group = null;

    public void create(PlayerData playerData) {
        JsonObject object = new JsonObject();

        object.add("name", playerData.name());
        object.add("group", default_group);
        object.add("prefix", permission.get(playerData.permission()).asObject().getString("prefix", "[orange]%1[orange] >[white] %2"));
        object.add("admin", playerData.isAdmin());

        permission_user.add(playerData.uuid(), object);
    }

    public void update() {
        for (JsonObject.Member p : permission_user) {
            JsonObject object = p.getValue().asObject();
            boolean isMatch = false;

            String group = p.getValue().asObject().get("group").asString();
            for (JsonObject.Member d : permission) {
                if (d.getName().equals(group)) {
                    isMatch = true;
                    break;
                }
            }

            if (!isMatch) permission_user.get(p.getName()).asObject().set("group", default_group);

            Player player = playerGroup.find(pl -> pl.uuid.equals(p.getName()));
            if (player != null && !p.getValue().asObject().get("name").asString().equals(player.name)) {
                player.name = object.getString("name", player.name);
            }
        }
        saveAll();
    }

    public void rename(String uuid, String new_name) {
        for (JsonObject.Member j : permission_user) {
            if (j.getName().equals(uuid)) {
                JsonObject o = j.getValue().asObject().set("name", new_name);
                o.set("name", new_name);
                permission_user.set(uuid, o);
                break;
            }
        }
    }

    public void saveAll() {
        root.child("permission_user.hjson").writeString(permission_user.toString(Stringify.FORMATTED));
    }

    public void reload(boolean init) {
        if (root.child("permission.hjson").exists()) {
            try {
                permission = JsonValue.readHjson(root.child("permission.hjson").reader()).asObject();
                for (JsonObject.Member data : permission) {
                    String name = data.getName();
                    if (permission.get(name).asObject().get("default") != null) {
                        if (default_group == null) {
                            if (permission.get(name).asObject().get("default").asBoolean()) {
                                default_group = name;
                            }
                        } else if (init) {
                            throw new PluginException(new Bundle(config.locale).get("system.perm.duplicate"));
                        }
                    }
                    if (permission.get(name).asObject().get("inheritance") != null) {
                        String inheritance = permission.get(name).asObject().getString("inheritance", null);
                        while (inheritance != null) {
                            for (int a = 0; a < permission.get(inheritance).asObject().get("permission").asArray().size(); a++) {
                                permission.get(name).asObject().get("permission").asArray().add(permission.get(inheritance).asObject().get("permission").asArray().get(a).asString());
                            }
                            inheritance = permission.get(inheritance).asObject().getString("inheritance", null);
                        }
                    }
                }

                if (default_group == null) {
                    for (JsonObject.Member data : permission) {
                        String name = data.getName();
                        if (name.equals("default")) {
                            default_group = name;
                            JsonObject json = readHjson(root.child("permission.hjson").reader()).asObject();
                            JsonArray perms = json.get("default").asObject().get("permission").asArray();
                            json.get("default").asObject().remove("permission");
                            json.get("default").asObject().add("default", true);
                            json.get("default").asObject().add("permission", perms);
                            root.child("permission.hjson").writeString(json.toString(Stringify.HJSON));
                        }
                    }
                }

                if (default_group == null) {
                    throw new PluginException(new Bundle(config.locale).get("system.perm.no-default"));
                }
            } catch (IOException e) {
                Log.err(e.getMessage());
                Core.app.dispose();
                Core.app.exit();
            } catch (Exception e) {
                new CrashReport(e);
            }
        } else {
            Log.warn("system.file-not-found", "permission.hjson");
        }

        if (root.child("permission_user.hjson").exists()) {
            try {
                permission_user = JsonValue.readHjson(root.child("permission_user.hjson").reader()).asObject();
                for (Player p : playerGroup.all()) {
                    p.isAdmin = isAdmin(vars.playerData().find(d -> d.name().equals(p.name)));
                }
            } catch (IOException | ParseException e) {
                // 이것도 유저들이 알아야 고침
                LoggerFactory.getLogger(Permission.class).error("Permission parsing", e);
            }
        } else {
            root.child("permission_user.hjson").writeString(new JsonObject().toString(Stringify.FORMATTED));
        }
    }

    public boolean check(Player player, String command) {
        PlayerData p = playerDB.get(player.uuid);

        if (!p.error()) {
            JsonValue object = permission_user.get(player.uuid);
            if (object != null) {
                JsonObject obj = object.asObject();
                int size = permission.get(obj.get("group").asString()).asObject().get("permission").asArray().size();
                for (int a = 0; a < size; a++) {
                    String permlevel = permission.get(obj.get("group").asString()).asObject().get("permission").asArray().get(a).asString();
                    if (permlevel.equals(command) || permlevel.equals("ALL")) {
                        return true;
                    }
                }
            } else {
                return false;
            }
        }
        return false;
    }

    public boolean isAdmin(PlayerData player) {
        try {
            if (permission_user.has(player.uuid())) {
                return permission_user.get(player.uuid()).asObject().getBoolean("admin", false);
            } else {
                return false;
            }
        } catch (NullPointerException ignored) {
            return false;
        }
    }

    public void setPermission_user(String old, String newid) {
        if (!old.equals(newid)) {
            JsonObject oldJson = permission_user.get(old).asObject();
            permission_user.set(newid, oldJson);
            permission_user.remove(old);
            update();
        }
    }
}

File: src/main/kotlin/remake/Permission.kt
package remake

import arc.Core
import arc.files.Fi
import mindustry.Vars.netServer
import mindustry.gen.Playerc
import org.hjson.*
import remake.Main.Companion.database
import java.util.*

object Permission {
    var perm = JsonObject()
    var data = JsonArray()
    var default = if (Config.authType == Config.AuthType.None) "user" else "visitor"
    private val root: Fi = Core.settings.dataDirectory.child("mods/Essentials/permission.txt")
    private val user: Fi = Core.settings.dataDirectory.child("mods/Essentials/permission_user.txt")

    val bundle = Bundle(Locale(System.getProperty("user.language"), System.getProperty("user.country")).toLanguageTag())

    val comment = """
        ${bundle["permission.sort"]}
        ${bundle["permission.notice"]}
        Usage
        {
            uuid: ${bundle["permission.usage.uuid"]}
            name: ${bundle["permission.usage.name"]}
            group: ${bundle["permission.usage.group"]}
            chatFormat: ${bundle["permission.usage.chatformat"]}
            admin: ${bundle["permission.usage.admin"]}
        }
        
        Examples
        [
            {
                uuid: uuids
                name: "my fun name"
            },
            {
                uuid: uuida
                chatFormat: "%1: %2"
            },
            {
                uuid: uuid123
                name: asdfg
                group: admin
                chatFormat: "[blue][ADMIN][]%1[white]: %2"
                admin: true
            }
        ]""".trimIndent()

    init {
        if (!root.exists()) {
            val json = JsonObject()

            val owner = JsonObject()
            owner.add("admin", true)
            owner.add("chatFormat", "[sky][Owner] %1[orange] > [white]%2")
            owner.add("permission", JsonArray().add("all"))

            val admin = JsonObject()
            val adminPerm = JsonArray()
            adminPerm.add("color")
            adminPerm.add("kill")
            adminPerm.add("mute")
            adminPerm.add("spawn")
            adminPerm.add("team")
            adminPerm.add("team.other")
            adminPerm.add("weather")
            adminPerm.add("info.other")

            admin.add("inheritance", "user")
            admin.add("admin", true)
            admin.add("chatFormat", "[yellow][Admin] %1[orange] > [white]%2")
            admin.add("permission", adminPerm)

            val user = JsonObject()
            val userPerm = JsonArray()
            userPerm.add("*login")
            userPerm.add("*reg")
            userPerm.add("ch")
            userPerm.add("discord")
            userPerm.add("info")
            userPerm.add("maps")
            userPerm.add("me")
            userPerm.add("motd")
            userPerm.add("players")
            userPerm.add("status")
            userPerm.add("time")
            userPerm.add("tp")
            userPerm.add("vote")

            user.add("inheritance", "visitor")
            user.add("chatFormat", "%1[orange] > [white]%2")
            user.add("permission", userPerm)

            val visitor = JsonObject()
            val visitorPerm = JsonArray()
            visitorPerm.add("help")
            visitorPerm.add("login")
            visitorPerm.add("reg")
            visitorPerm.add("t")

            visitor.add("chatFormat", "%1[scarlet] > [white]%2")
            visitor.add("default", true)
            visitor.add("permission", visitorPerm)

            json.add("owner", owner)
            json.add("admin", admin)
            json.add("user", user)
            json.add("visitor", visitor)

            root.writeString(json.toString(Stringify.HJSON))
        }

        if (!user.exists()) {
            val obj = JsonArray()
            obj.setComment(comment)
            Core.settings.dataDirectory.child("mods/Essentials/permission_user.txt").writeString(obj.toString(Stringify.HJSON_COMMENTS))
        }
    }

    fun save() {
        root.writeString(perm.toString(Stringify.HJSON))
    }

    fun sort() {
        user.writeString(data.setComment(comment).toString(Stringify.HJSON_COMMENTS))
    }

    @Throws(ParseException::class)
    fun load() {
        perm = JsonValue.readHjson(root.reader()).asObject()
        data = JsonValue.readHjson(user.reader()).asArray()

        for (data in perm) {
            val name = data.name
            if (Config.authType == Config.AuthType.None && perm.get(name).asObject().has("default")) {
                default = name
            }

            if (perm.get(name).asObject().has("inheritance")) {
                var inheritance = perm.get(name).asObject().getString("inheritance", null)
                while (inheritance != null) {
                    for (a in 0 until perm.get(inheritance).asObject()["permission"].asArray().size()) {
                        if (!perm.get(inheritance).asObject()["permission"].asArray()[a].asString().contains("*")) {
                            perm.get(name).asObject().get("permission").asArray().add(perm.get(inheritance).asObject()["permission"].asArray()[a].asString())
                        }
                    }
                    inheritance = perm.get(inheritance).asObject().getString("inheritance", null)
                }
            }
        }

        for (a in JsonValue.readHjson(user.reader()).asArray()) {
            val b = a.asObject()
            val c = database.players.find { e -> e.uuid == b.get("uuid").asString() }
            if (c == null){
                val data = database[b.get("uuid").asString()]
                if (data != null && b.has("group")) {
                    data.permission = b.get("group").asString()
                    database.update(b.get("uuid").asString(), data)
                }
            } else {
                c.permission = b.get("group").asString()
            }
        }
    }

    operator fun get(player: Playerc): PermissionData {
        val result = PermissionData
        val p = database.players.find { e -> e.uuid == player.uuid() }

        result.uuid = player.uuid()
        result.name = netServer.admins.findByIP(player.ip()).lastName
        result.group = p?.permission ?: default
        result.chatFormat = PermissionData.chatFormat
        result.admin = false

        data.forEach {
            val data = it.asObject()

            if (data.has("uuid") && data.get("uuid").asString().equals(player.uuid())) {
                result.uuid = data.getString("uuid", player.uuid())
                result.name = data.getString("name", netServer.admins.findByIP(player.ip()).lastName)
                result.group = data.getString("group", default)
                result.chatFormat = data.getString("chatFormat", PermissionData.chatFormat)
                result.admin = data.getBoolean("admin", false)
            }

            return result
        }
        return result
    }

    fun check(player: Playerc, command: String): Boolean {
        val data = get(player).group
        val size = perm[data].asObject()["permission"].asArray().size()
        for (a in 0 until size) {
            val node = perm[data].asObject()["permission"].asArray()[a].asString()
            if (node == command || node.equals("all", true)) {
                return true
            }
        }
        return false
    }

    object PermissionData {
        var name = ""
        var uuid = ""
        var group = default
        var chatFormat = "%1[orange] > [white]%2"
        var admin = false
    }
}